engine/src/main/java/io/camunda/zeebe/engine/processing/message/MessageTimeToLiveChecker.java

```diff
@@ -8,35 +8,89 @@
 import io.camunda.zeebe.stream.api.scheduling.TaskResult;
 import io.camunda.zeebe.stream.api.scheduling.TaskResultBuilder;
+import java.time.Duration;
+import org.agrona.collections.MutableInteger;
 
+/**
+ * The Message TTL Checker looks for expired message deadlines, and for each of those it writes an
+ * EXPIRE Message command.
+ *
+ * <p>To prevent that it clogs the log stream with too many EXPIRE Message commands, it only writes
+ * a limited number of these commands in a single run of {@link #execute(TaskResultBuilder)}.
+ *
+ * <p>It determines whether to reschedule itself immediately, or after the configured {@link
+ * MessageObserver#MESSAGE_TIME_TO_LIVE_CHECK_INTERVAL interval}. If it reschedules itself
+ * immediately, then it will continue where it left off the last time. Otherwise, it starts with the
+ * first expired message deadline it can find.
+ */
 public final class MessageTimeToLiveChecker implements Task {
 
   private static final MessageRecord EMPTY_DELETE_MESSAGE_COMMAND =
       new MessageRecord().setName("").setCorrelationKey("").setTimeToLive(-1L);
 
+  private static final int EXPIRE_COMMANDS_LIMIT_DEFAULT = 10;
+
+  /** This determines the maximum number of EXPIRE commands it will attempt to fit in the result. */
+  private static final int LIMIT = EXPIRE_COMMANDS_LIMIT_DEFAULT;
+
+  private final ProcessingScheduleService scheduleService;
   private final MessageState messageState;
 
-  public MessageTimeToLiveChecker(final MessageState messageState) {
+  /** Keeps track of the timestamp to compare the message deadlines against. */
+  private long currentTimestamp = -1;
+  /** Keeps track of where to continue between iterations. */
+  private MessageState.Index lastIndex;
+
+  public MessageTimeToLiveChecker(
+      final ProcessingScheduleService scheduleService, final MessageState messageState) {
+    this.scheduleService = scheduleService;
     this.messageState = messageState;
+    lastIndex = null;
   }
 
   @Override
   public TaskResult execute(final TaskResultBuilder taskResultBuilder) {
-    messageState.visitMessagesWithDeadlineBefore(
-        ActorClock.currentTimeMillis(),
-        expiredMessageKey -> writeDeleteMessageCommand(expiredMessageKey, taskResultBuilder));
-    return taskResultBuilder.build();
-  }
+    if (currentTimestamp == -1) {
+      currentTimestamp = ActorClock.currentTimeMillis();
+    }
 
-  private boolean writeDeleteMessageCommand(
-      final long expiredMessageKey, final TaskResultBuilder taskResultBuilder) {
-    return taskResultBuilder.appendCommandRecord(
-        expiredMessageKey, MessageIntent.EXPIRE, EMPTY_DELETE_MESSAGE_COMMAND);
+    final var counter = new MutableInteger(0);
+    final boolean shouldContinueWhereLeftOff =
+        messageState.visitMessagesWithDeadlineBeforeTimestamp(
+            currentTimestamp,
+            lastIndex,
+            (deadline, expiredMessageKey) -> {
+              final var newIndex = new Index(expiredMessageKey, deadline);
```

Comment:
Yes, the order in which these are passed into the Index record is irrelevant to the execution. The order is only relevant for the key in the column family because that is ordered. This index is just a convenient way to track uniquely identifying data.

Reply:
Yep saw it later, missed to remove the comment.